### PR TITLE
Per-container CPU usage in /proc/stat

### PR DESCRIPTION
**Pull request by aither64**
As it is, CPU usage stats in `/proc/stat` are shared with the host and all containers (issue #180). I'm trying to find a way for each container to see only its own CPU usage.

With this patch, `proc_stat_read()` reads per-container CPU usage from the *cpuacct* cgroup, `cpuacct.usage_all` to be exact. However, this file contains only usage in `user` and `system` modes, other fields from `/proc/stat` are not present. Idle time can be calculated, but as for other values like `nice` or `iowait`, they're all reported as zeroes in the resulting `/proc/stat` file. I don't think this can be solved without the kernel providing more information.

*cpuacct* cgroup v1 has to be present, otherwise `proc_stat_read()` will return an error. This could be changed if needed, I don't know if LXCFS is supposed to work without this cgroup somewhere.

We've been using it for several weeks on a staging node with [vpsAdminOS](https://github.com/vpsfreecz/vpsadminos) and it seems to work as expected.

### CODE REVIEW

**Comment by brauner** [comment]
How does this deal with the case where the `cpuacct` controller is missing. It seems to me that we should just return the host values in this case and not fail/error.

**Comment by aither64** [comment]
Alright, it's going to require more code and testing though, I haven't thought about it before. I will let you know when I have something, next week or so.

**Comment by aither64** [comment]
That is, if otherwise this patch is fine with you -- those zeroes instead of many fields in `/proc/stat`, I can't do anything about that afaik.

**Comment by brauner** [comment]
@aither64, it might be acceptable. I have no immediate objections to it. @hallyn will need to yay or nay as well though. :)

**Comment by hallyn** [comment]
zeroes where we can't get real values is fine, I think.  (An error when cpuacct.v1 is not available is not).

thanks.

**Comment by aither64** [comment]
Thank you for the info.

I've updated the patch. It now reads from `cpuacct` only if it is available, otherwise it falls back to the host's `/proc/stat`, just as it worked before. I've tested it in my environment and it looks good. It would be great if someone else could test this as well.

One additional imperfection that I've noticed is that CPU idle time is still leaking from the host, i.e. when you start a new container, `user` and `system` counters start from zero, but `idle` is based on the host's current value. I don't think it's a big issue, since I imagine tools working on usage data from `/proc/stat` are looking for relative changes of those counters and not the absolute values.

@brauner, @hallyn, please check it out. Thanks!

File: bindings.c [hunk 1]
```diff
@@ -3794,6 +3799,90 @@ static uint64_t get_reaper_age(pid_t pid)
...
+	struct cpuacct_usage *cpu_usage;
+	int rv = 0, i, j, ret, read_pos = 0, read_cnt;
+	int cg_cpu;
+	uint64_t cg_user, cg_system;
+	int64_t ticks_per_sec;
+	char *usage_str = NULL;
+
+	ticks_per_sec = sysconf(_SC_CLK_TCK);
+
+	if (ticks_per_sec < 0 && errno == EINVAL) {
```
Comment (brauner) [inline bindings.c]: Nit: unnecessary new line :)

File: bindings.c [hunk 2]
```diff
@@ -3794,6 +3799,90 @@ static uint64_t get_reaper_age(pid_t pid)
...
 
+/*
+ * Returns 0 on success.
+ * It is the caller's responsibility to free `return_usage`, unless this
+ * function returns an error.
+ */
+static int read_cpuacct_usage_all(char *cg, char *cpuset, struct cpuacct_usage **return_usage)
+{
+
+	int cpucount = get_nprocs();
```
Comment (brauner) [inline bindings.c]: `get_nprocs()` always returns a positive value? If not the `malloc()` below will be might funny...

File: bindings.c [hunk 3]
```diff
@@ -3794,6 +3799,90 @@ static uint64_t get_reaper_age(pid_t pid)
...
+	struct cpuacct_usage *cpu_usage;
+	int rv = 0, i, j, ret, read_pos = 0, read_cnt;
```
Comment (brauner) [inline bindings.c]: Nit: If you do `int rv = -1` here you can spare all those `rv = -1` in your error paths and simply return and also remove `{`, `}`. :)

File: bindings.c [hunk 4]
```diff
@@ -3794,6 +3799,90 @@ static uint64_t get_reaper_age(pid_t pid)
...
+	read_pos += read_cnt;
+
+	for (i = 0, j = 0; i < cpucount; i++) {
+		ret = sscanf(usage_str + read_pos, "%d %lu %lu\n%n", &cg_cpu, &cg_user,
+				&cg_system, &read_cnt);
+
+		if (ret == EOF) {
+			break;
+
+		} else if (ret != 3) {
```
Comment (brauner) [inline bindings.c]: Nit: There's no need for the `if else` ladder here. You can just do:
```
if (ret == EOF)
        break;

if (ret != 3) {
.
.
.
```

File: bindings.c (hunk 2 above)
Comment (aither64) [reply bindings.c]: Manpage says: *The function get_nprocs() returns the number of processors currently available in the  system.* So I expect it should be always at least `1`.

File: bindings.c (hunk 3 above)
Comment (aither64) [reply bindings.c]: That's true, but I prefer to make the return value explicit, e.g. someone can change the default `rv` in the future and it won't break.

File: bindings.c (hunk 1 above)
Comment (aither64) [reply bindings.c]: Fixed

File: bindings.c (hunk 4 above)
Comment (aither64) [reply bindings.c]: Fixed

**Pull request merged**